File: feature-add-history/src/main/java/home/oleg/feature_add_history/interactor/CheckInOut.java
package home.oleg.feature_add_history.interactor;

import home.oleg.placenearme.models.DetailedVenue;
import home.oleg.placenearme.repositories.VenueHistoryRepository;
import io.reactivex.Single;

public class CheckInOut {

    private final VenueHistoryRepository venueHistoryRepository;

    public CheckInOut(VenueHistoryRepository venueHistoryRepository) {
        this.venueHistoryRepository = venueHistoryRepository;
    }

    public Single<Boolean> execute(DetailedVenue detailedVenue) {
        if (detailedVenue.isHereNow()) {
            return venueHistoryRepository.checkOut(detailedVenue.getId())
                    .andThen(Single.just(false));
        } else {
            return venueHistoryRepository
                    .checkOutFromCurrent()
                    .andThen(venueHistoryRepository.checkIn(detailedVenue))
                    .andThen(Single.just(true));
        }
    }

}

File: feature-add-history/src/main/java/home/oleg/feature_add_history/domain/interactor/CheckInOut.kt
package home.oleg.feature_add_history.domain.interactor

import home.oleg.placenearme.models.DetailedVenue
import home.oleg.placenearme.repositories.VenueHistoryRepository
import io.reactivex.Single
import javax.inject.Inject

class CheckInOut @Inject constructor(private val venueHistoryRepository: VenueHistoryRepository) {

    fun execute(detailedVenue: DetailedVenue): Single<Boolean> {
        return if (detailedVenue.isHereNow) {
            venueHistoryRepository.checkOut(detailedVenue.id)
                    .andThen(Single.just(false))
        } else {
            venueHistoryRepository
                    .checkOutFromCurrent()
                    .andThen(venueHistoryRepository.checkIn(detailedVenue))
                    .andThen(Single.just(true))
        }
    }

}
